scripts/compatibility_scanner/compatibility_scanner.q

```diff
@@ -0,0 +1,239 @@
+
+  args:(enlist[`]!enlist[::]),.Q.opt .z.x;
+
+  if[0h~type args`dir;args[`dir]:first args`dir];
+  if[10h<>type args`dir;args[`dir]:()];
+
+  if[0h<>type args`file;args[`file]:enlist args`file];
+  args[`file]:args[`file] where 10h=type each args`file;
+
+  if[0h<>type args`exclude;args[`exclude]:enlist args`exclude];
+  args[`exclude]:args[`exclude] where 10h=type each args`exclude;
+
+  args[`regex]:$[0<>count .z.x;"--regex" in .z.x;0b];
+
+  if[0h~type args`csv;args[`csv]:first args`csv];
+  if[10h<>type args`csv;args[`csv]:()];
+
+  :args;
+ };
+
+loadChecksDict:{[]
+  checksDict:enlist[`z_assignment]!enlist readAssignmentChecks ASSIGNMENT_CHECKS_TSV;
+  checksDict,:enlist[`system_command]!enlist readCommandsChecks COMMANDS_CHECKS_TSV;
+
+  if[DEBUG_SHOW_REGEX_PASSED;
+    -1"DEBUG: Regex pattern being passed:";
+    -1 combineRegexList value checksDict;
+  ];
+  
+  :checksDict;
+ };
+
+readCommandsChecks:{[file]
+  tbl:("**";enlist"\t") 0: hsym`$file;
+
+  :combineRegexList tbl`regex;
+ };
+
+readAssignmentChecks:{[file]
+  tbl:("**";enlist"\t") 0: hsym`$file;
+
+  :combineRegexList raze{
+    :enlist[y] cross x cross enlist z;
+  }[ZS_TO_CHECK]'[tbl`prefix;tbl`suffix];
+ };
+
+combineRegexList:{[checksList]
+  :-1 _ raze{"(",x,")|"}each checksList;
+ };
+
+filterExcluded:{[files;excludedFiles]
+  :files where not files in excludedFiles;
+ };
+
+writeToCsv:{[file;checksDict;lines]
+  file:getFullPath file;
+  if[not file like "*[.]csv";
+    -1"WARN: Adding '.csv' suffix to file '",file,"'";
+    file,:".csv";
+  ];
+
+  tbl:categoriseLines[checksDict;lines];
+
+  -1"Saving to file '",file,"' . . .";
+  @[{(hsym`$x) 0: csv 0: y;-1"Saved"}[file];tbl;
+    {[x;y] -2"ERROR: Could not save to '",x,"' due to:\n",y}[file]];
+ };
+
+categoriseLines:{[checksDict;lines]
+  -1"Categorising lines . . .";
+
+  tbl:raze categoriseLine[checksDict]each lines;
+
+  -1"Categorised lines";
+
+  :tbl;
+ };
+
+categoriseLine:{[checkDict;line]
+  .progressTracker.dotNum:@[{mod[1+value x;4]};`.progressTracker.dotNum;0];
+  1"Categorising ",.progressTracker.dotNum#".";  // Used to give the user a visual indication that the program is still progressing without clogging up the logs
+
+  res:raze{[check;checkRegex;line]
+    lineDict:{`file`lineNum`code!(x 0;"J"$x 1;":" sv 2 _ x)}":" vs line;
+
+    found:0<"J"$first system"printf '%s' $(echo '",ssr[lineDict[`code];"'";"'\\''"],"' | grep -cE '",checkRegex,"' )";
+    if[not found;:()];
+
+    :([]
+      category:enlist check;
+      file:enlist lineDict`file;
+      line_num:enlist lineDict`lineNum;
+      code:enlist lineDict`code
+    );
+  }[;;line]'[key checkDict;value checkDict];
+
+  1"\033[2K\r";  // Clearing the line for the next step to print over it (ANSI escape code to erase current line + carriage return)
+
+  :res;
+ };
+
+showHelp:{[]
+  -1"+--------------------------------+";
+  -1"| FinSpace Compatibility Scanner |";
+  -1"+--------------------------------+";
+  -1"[Arguments]";
```

Comment:
Can you give an example input? Also some kind of indication of if the arguments are required or optional